dali/operators/math/expressions/broadcasting.cc

```diff
@@ -0,0 +1,252 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/math/expressions/broadcasting.h"
+#include <string>
+#include <vector>
+#include "dali/operators/math/expressions/arithmetic_meta.h"
+
+namespace dali {
+
+void CheckNumSamples(span<const TensorListShape<>*> shapes) {
+  if (shapes.size() <= 1)
+    return;
+  int nsamples = shapes[0]->num_samples();
+  for (int i = 1; i < shapes.size(); i++) {
+    DALI_ENFORCE(nsamples == shapes[i]->num_samples(),
+                 make_string("Number of samples should match. Got shapes with ", nsamples, " and ",
+                             shapes[i]->num_samples(), " respectively."));
+  }
+}
+
+template <typename Shape>
+int BroadcastNdimImpl(span<const Shape*> shapes) {
+  DALI_ENFORCE(shapes.size() >= 1);
+  int ndim = shapes[0]->sample_dim();
+  for (int i = 1; i < shapes.size(); i++) {
+    ndim = std::max(ndim, shapes[i]->sample_dim());
+  }
+  return ndim;
+}
+
+int BroadcastNdim(span<const TensorShape<>*> shapes) {
+  return BroadcastNdimImpl(shapes);
+}
+
+int BroadcastNdim(span<const TensorListShape<>*> shapes) {
+  return BroadcastNdimImpl(shapes);
+}
+
+/**
+ * @brief Implementation helper for `BroadcastShape` and `CanBroadcast`
+ *        The function updates val with the d-th dimension of `sh`, or returns false.
+ *        `val` can be updated with a new value if its current value is 1 or equal to the new one.
+ * @tparam ShapeLike e.g. span<const int64_t>, TensorShape<>
+ * @param val extent value to update (in an output shape)
+ * @param sh one of input shape
+ * @param d one of the dimension indices
+ * @return true if val could be updated (broadcastable)
+ * @return false if val could not be updated (non broadcastable)
+ */
+template <typename ShapeLike>
+bool BroadcastShapeImpl(int64_t& val, const ShapeLike& sh, int d) {
```

Comment:
I think some name like `TryBroadcastDim` would be better:
```suggestion
bool TryBroadcastDim(int64_t& target_value, const ShapeLike& shape_to_bcast, int dimension) {
```
or something along the lines.